src/hotspot/share/jeandle/jeandleCallVM.cpp

```diff
@@ -92,5 +92,18 @@ void JeandleCallVM::generate_call_VM(const char* name, address c_func, llvm::Fun
   ir_builder.CreateStore(ir_builder.getInt64((intptr_t)nullptr), last_Java_pc_ptr);
 
   // Return.
-  ir_builder.CreateRetVoid();
+  if (func_type->getReturnType()->isVoidTy()) {
+    ir_builder.CreateRetVoid();
+  } else {
+      llvm::Value* ret_val = call_c_func;
+      
+      // If the return type is a pointer type (object), we need to load from vm_result field
+      if (func_type->getReturnType()->isPointerTy()) {
```

Comment:
Also check addrspace of the return pointer. Only `llvm::jeandle::AddrSpace::JavaHeapAddrSpace` represent java objects.